README
Complete documentation for CL-FAD can be found in the 'doc'
directory.

CL-FAD also supports Nikodemus Siivola's HYPERDOC, see
<http://common-lisp.net/project/hyperdoc/> and
<http://www.cliki.net/hyperdoc>.

1. Installation

1.1. Probably the easiest way is

       (load "/path/to/cl-fad/load.lisp")

     This should compile and load CL-FAD on most Common Lisp
     implementations.

1.2. With MK:DEFSYSTEM you can make a symbolic link from
     'cl-fad.system' and 'cl-fad-test.system' to your central registry
     (which by default is in '/usr/local/lisp/Registry/') and then issue
     the command

       (mk:compile-system "cl-fad")

     Note that this relies on TRUENAME returning the original file a
     symbolic link is pointing to. This will only work with AllegroCL
     6.2 if you've applied all patches with (SYS:UPDATE-ALLEGRO).

1.3. You can also use ASDF instead of MK:DEFSYSTEM in a similar way
     (use the .asd files instead of the .system files).

1.4. Via QuickLisp:

	   (ql:quickload :cl-fad)

2. Test

CL-FAD comes with a small test suite.  To start it just load the file
"test.lisp" and evaluate (CL-FAD-TEST:TEST).


CHANGELOG
Version 0.7.2
2013-07-03
Fix documentation glitch (inconsistent download link) (Luís Oliveira)

Version 0.7.1
2013-02-18
Fix for LispWorks (R. Wilker)
Add :description to .asd file

Version 0.7.0
2013-01-23
Tests, pathname manipulation functions (Marco Baringer)
Temporary files (merged by Marco Baringer)
Fix symlink behaviour for some platforms (Mihai Bazon and Janis Dzerins)

Version 0.6.4
2010-11-18
Adapt to newer ClozureCL version (patch from Zach Beane, thanks to Chun Tian and Ralph Moritz as well)

Version 0.6.3
2009-09-30
Removed dependency on :SB-EXECUTABLE (thanks to Attila Lendvai and Tobias Rittweiler)

Version 0.6.2
2008-03-12
Never version of OpenMCL have %RMDIR (thanks to Dmitri Hrapof)

Version 0.6.1
2007-12-29
Integrated CLISP patch for LIST-DIRECTORY sent by Dan Muller

Version 0.6.0
2007-05-28
Support for Scieneer CL (patch from Douglas Crosher)

Version 0.5.2
2007-05-15
Fix for (newer versions of) ECL (patch from Dustin Long)

Version 0.5.1
2006-08-11
Added CHECKP to COPY-STREAM

Version 0.5.0
2006-04-21
Added :BREADTH-FIRST option to WALK-DIRECTORY (thanks to Mac Chan)

Version 0.4.3
2006-03-15
For CMUCL use TRUENAME with UNIX-RMDIR to cope with search lists (reported by Pawel Ostrowski)

Version 0.4.2
2006-01-04
WALK-DIRECTORY now catches circular symbolic links (thanks to Gary King)

Version 0.4.1
2006-01-03
Be more careful in DIRECTORY-WILDCARD (thanks to Gary King)
Patches for MCL (thanks to Gary King)

Version 0.4.0
2005-12-10
Exported COPY-STREAM (suggested by Chris Dean)

Version 0.3.3
2005-11-14
Fixed %RMDIR for newer versions of OpenMCL (thanks to James Bielman)

Version 0.3.2
2005-09-11
Fixed docs (correct name DELETE-DIRECTORY-AND-FILES)
Fixed docs (OVERWRITE was missing in COPY-FILE signature)
Added Debian link

Version 0.3.1
2005-06-02
Fixed typo in fad.lisp (thanks to Jack D. Unrue)

Version 0.3.0
2005-06-01
Support for ABCL (thanks to Jack D. Unrue)

Version 0.2.0
2005-05-29
Support for ECL (thanks to Maciek Pasternacki)

Version 0.1.3
2005-04-27
Changed implementation of DIRECTORY-EXISTS-P for LispWorks

Version 0.1.2
2005-03-17
Fixed typo in cl-fad.system (tanks to Andrew Philpot)

Version 0.1.1
2005-01-22
Fixed typos and versioning

Version 0.1.0
2005-01-22
Initial release


test.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-FAD-TEST; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-fad/test.lisp,v 1.12 2009/09/30 14:23:10 edi Exp $

;;; Copyright (c) 2004-2010, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:cl-fad-test)

(defparameter *tmp-dir*
              #+(or :win32 :mswindows :windows) "c:\\tmp\\"
              #-(or :win32 :mswindows :windows) "/tmp/")

(defvar *test-counter* 0)

(defmacro assert* (form)
  `(progn
     (format t "Trying to assert ~A~%" ',form)
     (assert ,form)
     (format t "Test ~A passed.~%" (incf *test-counter*))))

(defun test ()
  (setq *test-counter* 0)
  (let ((fad-dir (merge-pathnames (pathname-as-directory "fad-test")
                                  *tmp-dir*)))
    (delete-directory-and-files fad-dir :if-does-not-exist :ignore)
    (assert* (directory-pathname-p fad-dir))
    (assert* (directory-pathname-p (pathname *tmp-dir*)))
    (let ((foo-file (merge-pathnames "foo.lisp"
                                     fad-dir)))
      (assert* (not (directory-pathname-p foo-file)))
      (assert* (not (file-exists-p foo-file)))
      (assert* (not (file-exists-p fad-dir)))
      (with-open-file (out (ensure-directories-exist foo-file)
                           :direction :output
                           :if-does-not-exist :create)
        (write-string "NIL" out))
      (assert* (file-exists-p foo-file))
      (assert* (not (directory-exists-p foo-file)))
      (assert* (file-exists-p fad-dir))
      (assert* (directory-exists-p fad-dir))
      (assert* (equal fad-dir
                      (pathname-as-directory fad-dir)))
      (assert* (equal foo-file
                      (pathname-as-file foo-file)))
      (assert* (not (equal fad-dir
                           (pathname-as-file fad-dir))))
      (assert* (not (equal foo-file
                           (pathname-as-directory foo-file))))
      (dolist (name '("bar" "baz"))
        (let ((dir (merge-pathnames (pathname-as-directory name)
                                    fad-dir)))
          (dolist (name '("foo.text" "bar.lisp"))
            (let ((file (merge-pathnames name dir)))
              (with-open-file (out (ensure-directories-exist file)
                                   :direction :output
                                   :if-does-not-exist :create)
                (write-string "NIL" out))))))
      ;; /tmp/fad-test/foo.lisp
      ;; /tmp/fad-test/bar/bar.lisp
      ;; /tmp/fad-test/bar/foo.text
      ;; /tmp/fad-test/baz/bar.lisp
      ;; /tmp/fad-test/baz/foo.text
      ;; files : 5
      ;; dirs : 3
      (let ((file-counter 0)
            (file-and-dir-counter 0)
            (bar-counter 0))
        (walk-directory fad-dir
                        (lambda (file)
                          (declare (ignore file))
                          (incf file-counter)))
        ;; file-counter => 5
        (walk-directory fad-dir
                        (lambda (file)
                          (declare (ignore file))
                          (incf file-and-dir-counter))
                        :directories t)
        ;; file-and-dir-counter => 5 + 3
        (walk-directory fad-dir
                        (lambda (file)
                          (declare (ignore file))
                          (incf bar-counter))
                        :test (lambda (file)
                                (string= (pathname-name file)
                                         "bar"))
                        :directories t)
        ;; do not traverse the baz directory
        (walk-directory fad-dir
                        (lambda (file)
                          (declare (ignore file))
                          (incf file-and-dir-counter))
                        :test (lambda (file)
                                (not (and (directory-pathname-p file)
                                          (string= (first (last (pathname-directory file)))
                                                   "baz"))))
                        :directories :breadth-first)
        ;; file-and-dir-counter => 5 + 3 + 2 dirs + 3 files
        (assert* (= 5 file-counter))
        (assert* (= 13 file-and-dir-counter))
        (assert* (= 2 bar-counter)))
      (let ((bar-file (merge-pathnames "bar.lisp" fad-dir)))
        (copy-file foo-file bar-file)
        (assert* (file-exists-p bar-file))
        (with-open-file (foo-stream foo-file :element-type '(unsigned-byte 8))
          (with-open-file (bar-stream bar-file :element-type '(unsigned-byte 8))
            (assert* (= (file-length foo-stream)
                        (file-length bar-stream)))
            (loop for foo-byte = (read-byte foo-stream nil nil)
                  for bar-byte = (read-byte bar-stream nil nil)
                  while (and foo-byte bar-byte)
                  do (assert* (eql foo-byte bar-byte))))))
      (let ((baz-dir (merge-pathnames (pathname-as-directory "baz")
                                      fad-dir))
            (list (mapcar #'namestring (list-directory fad-dir))))
        (assert* (find (namestring (truename foo-file)) list :test #'string=))
        (assert* (find (namestring (truename baz-dir)) list :test #'string=))
        (assert* (not (find (namestring (pathname-as-file baz-dir))
                            list
                            :test #'string=)))))
    (delete-directory-and-files fad-dir :if-does-not-exist :error)
    (assert* (not (file-exists-p fad-dir)))
    (assert* (not (directory-exists-p fad-dir))))
  (format t "All tests passed.~%"))


load.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-fad/load.lisp,v 1.9 2009/09/30 14:23:10 edi Exp $

;;; Copyright (c) 2004-2009, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :cl-user)

(defparameter *cl-fad-base-directory*
  (make-pathname :name nil :type nil :version nil
                 :defaults (parse-namestring *load-truename*)))

#+:allegro (require :osi)
#+:sbcl (require :sb-executable)
#+:sbcl (require :sb-posix)

(let ((cl-fad-base-directory
        (make-pathname :name nil :type nil :version nil
                       :defaults (parse-namestring *load-truename*))))
  (let (must-compile)
    #+:cormanlisp (declare (ignore must-compile))
    (dolist (file '("packages"
		    #+:cormanlisp "corman"
		    #+:openmcl "openmcl"
		    "fad"))
      (let ((pathname (make-pathname :name file :type "lisp" :version nil
                                     :defaults cl-fad-base-directory)))
        ;; don't use COMPILE-FILE in Corman Lisp, it's broken - LOAD
        ;; will yield compiled functions anyway
        #-:cormanlisp
        (let ((compiled-pathname (compile-file-pathname pathname)))
          (unless (and (not must-compile)
                       (probe-file compiled-pathname)
                       (< (file-write-date pathname)
                          (file-write-date compiled-pathname)))
            (setq must-compile t)
            (compile-file pathname))
          (setq pathname compiled-pathname))
        (load pathname)))))


path.lisp
(in-package :cl-fad)

(defmacro defalias (name args realname)
  `(progn
     (defun ,name ,args
       ,(if (eql '&rest (first args))
            `(apply #',realname ,(second args))
            `(,realname ,@args)))
     (define-compiler-macro ,name (&rest args)
       (list* ',realname args))))

(defalias path:dirname (pathname) cl-fad:pathname-directory-pathname)

(defalias path:basename (pathname) cl-fad:pathname-sans-directory-pathname)

(defalias path:-e (pathname) cl-fad:file-exists-p)

(defalias path:-d (directory) cl-fad:directory-exists-p)

(defalias path:catfile (&rest pathnames) cl-fad:merge-pathnames-as-file)

(defalias path:catdir (&rest pathnames) cl-fad:merge-pathnames-as-directory)

(defalias path:= (a b) cl-fad:pathname-equal)

(defalias path:absolute-p (pathname) cl-fad:pathname-absolute-p)

(defalias path:relative-p (pathname) cl-fad:pathname-relative-p)

(defalias path:root-p (pathname) cl-fad:pathname-root-p)

(defalias path:rm-r (pathname) cl-fad:delete-directory-and-files)


corman.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-fad/corman.lisp,v 1.5 2009/09/30 14:23:09 edi Exp $

;;; Copyright (c) 2004-2009, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHORS 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :cl)

(defun wild-pathname-p (pathspec &optional field)
  (unless (pathnamep pathspec)
    (setq pathspec (pathname pathspec)))
  (labels ((name-wild-p (name)
             (or (eq :wild name)
                 (and (stringp name)
                      (string= "*" name))))
           (dir-wild-p (dir)
             (or (find :wild dir)
                 (find :wild-inferiors dir)
                 (find "*" dir :test #'string=))))
    (case field
      ((:name)
       (name-wild-p (pathname-name pathspec)))
      ((:type)
       (name-wild-p (pathname-type pathspec)))
      ((:directory)
       (dir-wild-p (pathname-directory pathspec)))
      ((nil)
       (or (name-wild-p (pathname-name pathspec))
           (name-wild-p (pathname-type pathspec))
           (dir-wild-p (pathname-directory pathspec))))
      (t nil))))

(defun file-namestring (pathspec)
  (flet ((string-list-for-component (component)
           (cond ((eq component :wild)
                  (list "*"))
                 (component
                  (list component))
                 (t nil))))
    (let* ((pathname (pathname pathspec))
           (name (pathnames::pathname-internal-name pathname))
           (type (pathnames::pathname-internal-type pathname)))
      (format nil "~{~A~}~{.~A~}"
              (string-list-for-component name)
              (string-list-for-component type)))))

(in-package :win32)

(defwinapi RemoveDirectory
    ((lpPathName LPCSTR))
  :return-type BOOL
  :library-name "Kernel32"
  :entry-name "RemoveDirectoryA"
  :linkage-type :pascal)

(defun delete-directory (pathspec)
  "Deletes the empty directory denoted by the pathname designator
PATHSPEC.  Returns true if successful, NIL otherwise."
  (win:RemoveDirectory
   (ct:lisp-string-to-c-string
    (namestring (pathname pathspec)))))

(export 'delete-directory)


packages.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-fad/packages.lisp,v 1.12 2009/09/30 14:23:10 edi Exp $

;;; Copyright (c) 2004-2010, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:cl-user)

(defpackage :cl-fad
  (:nicknames :fad)
  (:use :cl)
  #+:allegro
  (:shadow :copy-file
           :delete-directory-and-files)
  #+:abcl
  (:shadow :list-directory)
  (:export :copy-file
           :copy-stream
           :delete-directory-and-files
           :directory-exists-p
           :directory-pathname-p
           :file-exists-p
           :list-directory
           :pathname-as-directory
           :pathname-as-file
           :pathname-directory-pathname
           :pathname-equal
           :pathname-parent-directory
           :pathname-absolute-p
           :pathname-relative-p
           :pathname-root-p
           
           :canonical-pathname
           :merge-pathnames-as-directory
           :merge-pathnames-as-file           
           
           :walk-directory

           :open-temporary
           :with-output-to-temporary-file
           :with-open-temporary-file
           :*default-template*
           :invalid-temporary-pathname-template
           :cannot-create-temporary-file
           #+win32 #:missing-temp-environment-variable))

(defpackage :path
  (:use)
  (:documentation "Rexporting certain functions from the cl-fad package with shorter names.

This package provides no functionality, it serves only to make file
system intensive code easier to read (for unix people at least).")
  (:export #:dirname
           #:basename
           #:-e
           #:-d
           #:catfile
           #:catdir
           #:rm-r
           #:=

           #:absolute-p
           #:relative-p
           #:root-p))

(defpackage :cl-fad-test
  (:use :cl :cl-fad :unit-test)
  (:export :test))


temporary-files.lisp
(in-package :cl-fad)

(defparameter *default-template* "TEMPORARY-FILES:TEMP-%")

(defparameter *max-tries* 10000)

(defvar *name-random-state* (make-random-state t))

;; from XCVB
(eval-when (:load-toplevel :execute)
  (defun getenv (x)
    "Query the libc runtime environment. See getenv(3)."
    (declare (ignorable x))
    #+(or abcl clisp xcl) (ext:getenv x)
    #+allegro (sys:getenv x)
    #+clozure (ccl:getenv x)
    #+(or cmu scl) (cdr (assoc x ext:*environment-list* :test #'string=))
    #+cormanlisp
    (let* ((buffer (ct:malloc 1))
           (cname (ct:lisp-string-to-c-string x))
           (needed-size (win:getenvironmentvariable cname buffer 0))
           (buffer1 (ct:malloc (1+ needed-size))))
      (prog1 (if (zerop (win:getenvironmentvariable cname buffer1 needed-size))
                 nil
                 (ct:c-string-to-lisp-string buffer1))
        (ct:free buffer)
        (ct:free buffer1)))
    #+ecl (si:getenv x)
    #+gcl (system:getenv x)
    #+lispworks (lispworks:environment-variable x)
    #+mcl (ccl:with-cstrs ((name x))
            (let ((value (_getenv name)))
              (unless (ccl:%null-ptr-p value)
                (ccl:%get-cstring value))))
    #+sbcl (sb-ext:posix-getenv x)
    #-(or abcl allegro clisp clozure cmu cormanlisp ecl gcl lispworks mcl sbcl scl xcl)
    (error "~S is not supported on your implementation" 'getenv))

  (defun directory-from-environment (environment-variable-name)
    (let ((string (getenv environment-variable-name)))
      (when (plusp (length string))
        (pathname-as-directory string))))

  #+win32
  (define-condition missing-temp-environment-variable (error)
    ()
    (:report (lambda (condition stream)
               (declare (ignore condition))
               (format stream "the TEMP environment variable has not been found, cannot continue"))))

  #+win32
  (defun get-default-temporary-directory ()
    (or (directory-from-environment "TEMP")
        (error 'missing-temp-environment-variable)))

  #-win32
  (defun get-default-temporary-directory ()
    (or (directory-from-environment "TMPDIR")
        #-clisp
        (probe-file #P"/tmp/")
        #+clisp
        (and (ext:probe-directory #P"/tmp/")
             #P"/tmp/")))

  (handler-case
      (logical-pathname-translations "TEMPORARY-FILES")
    (#-clisp type-error
     #+clisp simple-error ()
      (alexandria:if-let (default-temporary-directory (get-default-temporary-directory))
        (setf (logical-pathname-translations "TEMPORARY-FILES") `(("*.*.*" ,default-temporary-directory)))
        (warn "could not automatically determine a default mapping for TEMPORARY-FILES")))))

;; locking for multi-threaded operation with unsafe random function

(defvar *create-file-name-lock* (bordeaux-threads:make-lock "Temporary File Name Creation Lock"))

(defmacro with-file-name-lock-held (() &body body)
  `(bordeaux-threads:with-lock-held (*create-file-name-lock*)
     ,@body))

(defun generate-random-string ()
  (with-file-name-lock-held ()
    (format nil "~:@(~36,8,'0R~)" (random (expt 36 8) *name-random-state*))))

(define-condition invalid-temporary-pathname-template (error)
  ((string :initarg :string))
  (:report (lambda (condition stream)
             (with-slots (string) condition
               (format stream "invalid temporary file name template ~S, must contain a percent sign that is to be replaced by a random string" string)))))

(defun generate-random-pathname (template random-string-generator)
  (let ((percent-position (or (position #\% template)
                              (error 'invalid-temporary-pathname-template :string template))))
    (merge-pathnames (concatenate 'string
                                  (subseq template 0 percent-position)
                                  (funcall random-string-generator)
                                  (subseq template (1+ percent-position))))))

(define-condition cannot-create-temporary-file (error)
  ((template :initarg :template)
   (max-tries :initarg :max-tries))
  (:report (lambda (condition stream)
             (with-slots (template max-tries) condition
               (format stream "cannot create temporary file with template ~A, giving up after ~D attempt~:P"
                       template max-tries)))))

(defun open-temporary (&rest open-arguments
		       &key
                         (template *default-template*)
			 (generate-random-string 'generate-random-string)
                         (max-tries *max-tries*)
                         (direction :output)
			 &allow-other-keys)
  "Create a file with a randomly generated name and return the opened
   stream.  The resulting pathname is generated from TEMPLATE, which
   is a string representing a pathname template.  A percent sign (%)
   in that string is replaced by a randomly generated string to make
   the filename unique.  The default for TEMPLATE places temporary
   files in the TEMPORARY-FILES logical pathname host, which is
   automatically set up in a system specific manner.  The file name
   generated from TEMPLATE is merged with *DEFAULT-PATHNAME-DEFAULTS*,
   so random pathnames relative to that directory can be generated by
   not specifying a directory in TEMPLATE.

   GENERATE-RANDOM-STRING can be passed to override the default
   function that generates the random name component.  It should
   return a random string consisting of characters that are permitted
   in a pathname (logical or physical, depending on TEMPLATE).

   The name of the temporary file can be accessed calling the PATHNAME
   function on STREAM.  For convenience, the temporary file is opened
   on the physical pathname, i.e. if the TEMPLATE designate a logical
   pathname the translation to a physical pathname is performed before
   opening the stream.

   In order to create a unique file name, OPEN-TEMPORARY may loop
   internally up to MAX-TRIES times before giving up and signalling a
   CANNOT-CREATE-TEMPORARY-FILE condition."
  (loop thereis (apply #'open
                       (translate-logical-pathname (generate-random-pathname template generate-random-string))
                       :direction direction
                       :if-exists nil
                       (alexandria:remove-from-plist open-arguments :template :generate-random-string :max-tries))
        repeat max-tries
        finally (error 'cannot-create-temporary-file
                       :template template
                       :max-tries max-tries)))

(defmacro with-output-to-temporary-file ((stream &rest args) &body body)
  "Create a temporary file using OPEN-TEMPORARY with ARGS and run BODY
  with STREAM bound to the temporary file stream.  Returns the
  pathname of the file that has been created.  See OPEN-TEMPORARY for
  permitted options."
  `(with-open-stream (,stream (open-temporary ,@args))
     ,@body
     (pathname ,stream)))

(defmacro with-open-temporary-file ((stream &rest args &key keep &allow-other-keys) &body body)
  "Create a temporary file using OPEN-TEMPORARY with ARGS and run BODY
  with STREAM bound to the temporary file stream.  Returns the values
  returned by BODY.  By default, the file is deleted when BODY is
  exited. If a true value is passed in KEEP, the file is not deleted
  when the body is exited.  See OPEN-TEMPORARY for more permitted
  options."
  `(with-open-stream (,stream (open-temporary ,@(alexandria:remove-from-plist args :keep)))
     #+sbcl
     (declare (sb-ext:muffle-conditions sb-ext:code-deletion-note))
     ,(if (and (constantp keep)
               keep)
          `(progn ,@body)
          `(unwind-protect
                (progn ,@body)
             (unless ,keep
               (close ,stream)
               (delete-file (pathname ,stream)))))))


openmcl.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CCL; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/cl-fad/openmcl.lisp,v 1.6 2009/09/30 14:23:10 edi Exp $

;;; Copyright (c) 2004-2010, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHORS 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package :cl-fad)

(eval-when (:compile-toplevel :load-toplevel :execute)
  (flet ((ccl-function-feature (symbol-name feature)
           (let ((symbol (find-symbol symbol-name :ccl)))
             (when (and symbol (fboundp symbol))
               (pushnew feature *features*)))))
    (ccl-function-feature "%RMDIR" :ccl-has-%rmdir)
    (ccl-function-feature "DELETE-DIRECTORY" :ccl-has-delete-directory)))

(defpackage :cl-fad-ccl
  (:use :cl)
  (:export delete-directory)
  (:import-from :ccl
                :%realpath
                :signal-file-error
                :native-translated-namestring
                :with-cstrs)
  #+ccl-has-%rmdir
  (:import-from :ccl :%rmdir)
  #+ccl-has-delete-directory
  (:import-from :ccl :delete-directory))

(in-package :cl-fad-ccl)

#-ccl-has-%rmdir
(defun %rmdir (name)
  (with-cstrs ((n name))
    (#_rmdir n)))

;;; ClozureCL 1.6 introduced ccl:delete-directory with semantics that
;;; are acceptably similar to this "legacy" definition.

#-ccl-has-delete-directory
(defun delete-directory (path)
  (let* ((namestring (native-translated-namestring path)))
    (when (%realpath namestring)
      (let* ((err (%rmdir namestring)))
        (or (eql 0 err) (signal-file-error err path))))))

